### PR TITLE
added PHPStan static analysis

### PR DESCRIPTION
**Pull request by soukicz**
- bug fix? no
- new feature? no
- BC break? no

### CODE REVIEW

File: .phpstan.neon [hunk 1]
```diff
@@ -0,0 +1,4 @@
+parameters:
+	ignoreErrors:
+		- '#Function finfo_[a-zA-Z0-9_]+ not found#'
+		- '#Call to an undefined method Nette\\Mail\\FallbackMailer::onFailure\(\)#'
```
Comment (JanTvrdik) [inline .phpstan.neon]: If use install phpstan/phpstan-nette extension this should not be necessary.

File: .travis.yml [hunk 2]
```diff
@@ -17,6 +17,7 @@ matrix:
 script:
     - vendor/bin/tester tests -s $coverageArgs
     - php temp/code-checker/src/code-checker.php --short-arrays --strict-types
+    - vendor/bin/phpstan analyse -c .phpstan.neon --level 0 src
```
Comment (JanTvrdik) [inline .travis.yml]: Level 0? That checks almost nothing.

**Comment by dg** [comment]
PHPStan cannot be dependency, it is external tools so it must be installed via `composer create-project phpstan/phpstan`.

**Comment by dg** [comment]
I think it is very premature to use PHPStan as dependency in Travis. We should wait for a stable version. Currently I am unable even to install it.

image

**Comment by hrach** [comment]
That's because it is not prepared to be a project instead of dev-dependency. If you use it as dev-depency, of course it doesn't require it's own dev-dependencies.

**Comment by JanTvrdik** [comment]
[Related](https://github.com/nextras/orm/issues/208#issuecomment-281130310)

> I though about it while ago and came to the conclusion that considering how Composers currently works, installing tools (such as phpstan) via dev-dependency of the project is the best solution.
> 
> Installing via `composer create-project` makes it much harder to update the tool later and if you want to install optional dependency (e.g. phpstan/phpstan-nette) of the tool, then using `composer create-project` is especially hard and inconvenient.

**Comment by dg** [comment]
> That's because it is not prepared to be a project instead of dev-dependency

It is not related. Simply it requires stability=dev.

> that considering how Composer currently works, installing tools (such as phpstan) via dev-dependency of the project is the best solution.

Composer must be improved, because installing as dev-dependency is not good-enough solution. 

For example, it completely prevents to use PHPStan for Nette, because it has a (more strict) dependencies on Nette, and therefore it affects which versions of other libraries are used.

Btw, there is no problem to install PHPStan as project. Problem is that PHPStan cannot analyse project which uses the same classes as PHPStan itself. I solved it in Updater via completely virtual reflection.

**Pull request closed without merging**